apps/frontend/src/auth.ts

```diff
@@ -14,17 +15,72 @@ export const { handle, signIn, signOut } = SvelteKitAuth({
 	trustHost: true,
 	callbacks: {
 		session: async ({ session, token }) => {
-			if ('accessToken' in token) {
-				return { ...session, accessToken: token.accessToken };
+			if ('access_token' in token) {
+				return { ...session, accessToken: token.access_token };
 			}
 			return session;
 		},
-		jwt({ token, account }) {
+		// Ref: https://docs.descope.com/getting-started/nextauth/functions
+		jwt: async ({ token, account }) => {
 			if (account) {
-				// Add access token if provided.
-				return { ...token, accessToken: account.access_token };
+				console.info('Initial login');
+
+				if (account.expires_in === undefined) throw Error('Account has no expiration set.');
+
+				return {
+					...token,
+					access_token: account.access_token,
+					expires_at: Math.floor(Date.now() / 1000 + account.expires_in),
+					refresh_token: account.refresh_token
+				};
+			} else {
+				if (typeof token.expires_at !== 'number') throw Error('Token has no expiration set.');
+
+				if (Date.now() < token.expires_at * 1000) {
+					console.debug('Returning valid token');
+
+					return token;
+				} else {
+					console.info('Token expired, renewing', Date.now(), token.expires_at * 1000);
+
+					try {
+						if (typeof token.refresh_token !== 'string')
+							throw new Error('Token has no refresh token.');
+
+						const clientId = env.AUTH_DESCOPE_ID;
+						const clientSecret = env.AUTH_DESCOPE_SECRET;
+
+						if (!clientId || !clientSecret) {
+							throw new Error('Missing Descope client credentials');
+						}
```

Comment:
I would consider this an unexpected (programming) error and would move it outside of the try/catch which is making the API requests. https://svelte.dev/docs/kit/errors#Unexpected-errors